### PR TITLE
WIP: Experiment to use LIEF

### PR DESCRIPTION
**Pull request by lkollar**
Testing out LIEF for #174.

This will fail CI at the moment because of the missing wheel for CentOS 5.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,2 +1,3 @@
 wheel == 0.31.1
 pyelftools >=0.24
+https://lief-project.github.io/packages/lief/lief-0.10.0.dev0-cp37-cp37m-manylinux1_x86_64.whl
```
Comment (figroc) [inline requirements.txt]: `lief` has been published in pypi. Why don't you install it from there?

Comment (lkollar) [reply requirements.txt]: I am, but I haven't yet pushed the latest changes. I just did now. Not that this is still a work in progress.

**Comment by figroc** [comment]
I have to build on manylinux2010. So I applied this patch and built the patched `auditwheel` on manylinux2010. When I invoked `auditwheel repair` with `lief`, the new `cp36-cp36m-manylinux2010_x86_64` was built but with the following error:
```
Traceback (most recent call last):
  File "/usr/local/bin/auditwheel", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.6/site-packages/auditwheel/main.py", line 50, in main
    rval = args.func(args, p)
  File "/usr/local/lib/python3.6/site-packages/auditwheel/main_repair.py", line 97, in execute
    analyzed_tag = analyze_wheel_abi(out_wheel).overall_tag
  File "/usr/local/lib/python3.6/site-packages/auditwheel/wheel_abi.py", line 191, in analyze_wheel_abi
    uses_PyFPE_jbuf) = get_wheel_elfdata(wheel_fn)
  File "/usr/local/lib/python3.6/site-packages/auditwheel/wheel_abi.py", line 58, in get_wheel_elfdata
    for key, value in elf_find_versioned_symbols(elf):
  File "/usr/local/lib/python3.6/site-packages/auditwheel/elfutils.py", line 47, in elf_find_versioned_symbols
    for verneed, verneed_iter in section.iter_versions():
  File "/usr/local/lib/python3.6/site-packages/elftools/elf/gnuversions.py", line 154, in iter_versions
    for verneed, vernaux in super(GNUVerNeedSection, self).iter_versions():
  File "/usr/local/lib/python3.6/site-packages/elftools/elf/gnuversions.py", line 116, in iter_versions
    count_field, str(entry)))
  File "/usr/local/lib/python3.6/site-packages/elftools/common/utils.py", line 77, in elf_assert
    _assert_with_exception(cond, msg, ELFError)
  File "/usr/local/lib/python3.6/site-packages/elftools/common/utils.py", line 109, in _assert_with_exception
    raise exception_type(msg)
elftools.common.exceptions.ELFError: Expected number of version auxiliary entries (vn_cnt) to be > 0for the following version entry: Container({'vn_version': 1162, 'vn_cnt': 0, 'vn_file': 0, 'vn_aux': 16, 'vn_next': 0})
```

On importing the package on `python:3.6-slim`, I got the error:
```
Inconsistency detected by ld.so: dl-version.c: 205: _dl_check_map_versions: Assertion `needed != NULL' failed!
```

You can checkout the corresponding [scripts](https://github.com/figroc/s2geometry/tree/develop/cross]) and [Dockerfile](https://github.com/figroc/dockcross/blob/master/Dockerfile) which I use. Hope it would be helpful.

**Comment by lkollar** [comment]
> Inconsistency detected by ld.so: dl-version.c: 205: _dl_check_map_versions: Assertion `needed != NULL' failed!

I came across the same issue while I was testing LIEF with auditwheel. I think this might be due to a bug in LIEF but I can't reproduce the issue right now. What would be most useful is a concrete reproducer including the exact LIEF version, vendored library name and version and the environment, so we can further investigate this. I am not able to dedicate much time to this at the moment but I am planning to follow up soon.

@figroc if you could upload the wheel somewhere (before invoking auditwheel on it) I can check if I can reproduce the issue.

**Comment by figroc** [comment]
@lkollar You can build the wheel with the command:
```
docker run --rm -v $(pwd):/work --entrypoint bash figroc/dockcross:manylinux2010 -c 'pip3.6 -v wheel s2geometry'
```
The non-audited wheel file `s2geometry-0.9.0-cp36-cp36m-linux_x86_64.whl` will be generated in `$(pwd)`.

**Comment by lkollar** [comment]
@figroc thanks, I managed to reproduce the issue with `libcrpyto.so`. Opened an issue in LIEF: https://github.com/lief-project/LIEF/issues/374.

File: auditwheel/patcher.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+class Lief(ElfPatcher):
+    # TODO make context manager, save only when done with everything?
+    def replace_needed(self, file_name, so_name, new_so_name):
+        elf = lief.parse(file_name)
+        for lib in elf.dynamic_entries:
+            if lib.tag == lief.ELF.DYNAMIC_TAGS.NEEDED and lib.name == so_name:
+                logger.info("Replacing needed library: %s with %s",
+                            lib.name, new_so_name)
+                lib.name = new_so_name
+                elf.write(file_name)
```
Comment (JanuszL) [inline auditwheel/patcher.py:78]: In my case I also needed to:
```
for sym in elf.symbols_version_requirement:
    if sym.name == so_name:
        sym.name = new_so_name
```

**Comment by lkollar** [comment]
Closing this in favour of https://github.com/pypa/auditwheel/pull/240 which retains the interface for ELF patching as discussed in #174.

**Pull request closed without merging**